main_loop_tf/main.py

```diff
@@ -417,26 +467,43 @@ def __run(build_model):
                         val_outs,
                         val_summary_ops[s],
                         val_reset_cm_op,
+                        smaller_n_gpus,
                         which_set=s)
 
-
+# Multi-gpu building:
+# In each graph some operation are duplicated so that at runtime
+# it's possible to choose which towers have to be executed. This
+# allows to solve the problem of multi-gpu fault in the case of
+# insufficiently large batches.
+# In this new version the returned operation (for example those for
+# summaries computing) are in the form: [[..], [..]], so that the
+# list of operations for the full and reduced number of gpus used
+# are in the first and second position respectively. In this way
+# it's possible to choose at runtime in main loop or validation
+# which operations of the graph have to be called with respect to
+# the batch size.
+# For what concern the placeholders they now come up as a list of
+# input placeholders associated to each tower that is directly fed
+# from the main loop, so the tf.split is removed and the split
+# is performed out of the graph.
+# Note: the duplicated function are the ones with an 's' as prefix
+# in the name.
+# TODO: double check if this way to operate is correct.
+# In the specific check if duplicating the operation could affect in
+# in some manner the learning task.
```

Comment:
Move this comment inside the function, as a docstring and change it (with proper indentation) to
```python
''' Build the multiGPU graph of computation

This function creates a copy of the computation graph on each GPU. The result of the computation of each GPU is stored in a "tower"
Note that thanks to the use of name_scopes and variable_scopes, calling this function multiple times does not create multiple copies of the *Ops* and of the *Variables* (respectively), but rather only adds the Ops that change from one call to the other and reuses the same Variables.

Furthermore, to accommodate for the case where the last minibatch of a set is smaller than batch_size, we create two versions of each operation (e.g., those that compute the summaries), the first that uses all the GPUs (default) and the second that only uses the subset that is possible to use in the reduced batch case. Each returned value is then a list of two ops. This allows to choose at runtime which operations of the graph to call, depending on the batch size.
'''
```